File: WINGS_AI_Bootcamp/Python/NumToWords1.py
SPACE = " "
AND = "and"
HUN = "hundred"
def convert2digits(k: int) -> str:
    upto20 = ["", "One", "Two"," Three","Four","Five","Six","Seven","Eight","Nine","Ten","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"]
    tens = ["","","Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]
    if k < 20:
        return upto20[k]
    else:
        return tens[k // 10] + SPACE + upto20[k % 10]

def convert3digits(k: int) -> str:
    if k < 100:
        return convert2digits(k)
    else:
        h, tu = divmod(k, 100)
        if tu == 0:
            return convert2digits(k // 100) + SPACE + HUN
        else:
            return convert2digits(h) + SPACE + HUN + SPACE + AND + SPACE + convert2digits(tu)
    
print(convert3digits(671))



WESTERN =  { 10 ** 12 : "Trillion",10 ** 9 : "Billion", 10 ** 6 : "Million", 10 ** 3 : "Thousand", 1: ""}
INDIAN = {10 ** 9: "Arab", 10 ** 7 : "Crore", 10 ** 5 : "Lakh", 1000: "Thousand", 1: ""}


def to_words(system: dict, amount: int) -> str:
    denoms = sorted(system.keys(), reverse = True)
    in_words = ""
    for denom in denoms:
        how_many, amount = divmod(amount, denom)
        if how_many > 0:
            in_words += convert3digits(how_many)+ SPACE + system[denom] + SPACE 
    return in_words

File: WINGS_AI_Bootcamp/Haskell/NumToWords1.hs
import qualified Data.Map as M
import Data.List (sort, reverse)

oneWords :: [String]
oneWords = ["", "One", "Two","Three", "Four", "Five","Six", "Seven", "Eight","Nine", "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"]

tenWords :: [String]
tenWords = ["", "", "Twenty", "Thirty", "Forty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"]

wSystem :: M.Map Int String
wSystem = M.fromList [(10^12, "Trillion"), 
                      (10^9, "Billion"), 
                      (10^6, "Million"), 
                      (1000, "Thousand"), 
                      (1, "")]

iSystem :: M.Map Int String  -- Fixed typo here
iSystem = M.fromList [(10^9, "Arab"), 
                      (10^7, "Crore"), 
                      (10^5, "Lakh"), 
                      (1000, "Thousand"), 
                      (1, "")]

breakInto' :: Int -> [Int] -> [(Int, Int)]
breakInto' _ []  = []
breakInto' amt [1] = [(amt, 1)]
breakInto' amt (d: ds) = (div amt d, d) : breakInto' (rem amt d) ds  -- call breakInto' recursively

breakInto :: Int -> [Int] -> [(Int, Int)]
breakInto amt ds = filter (\a -> fst a /= 0) $ breakInto' amt ds

convert :: Int -> String 
convert amt 
    | amt < 20 = oneWords !! amt
    | amt < 100 = tenWords !! (div amt 10) ++ " " ++ oneWords !! (rem amt 10)
    | rem amt 100 == 0 = oneWords !! (div amt 100) ++ " hundred"
    | otherwise = oneWords !! (div amt 100) ++ " Hundred and " ++ convert (rem amt 100)

main = do
    print $ M.keys iSystem
    print $ breakInto' 167234520 $ reverse $ sort $ M.keys wSystem
    print $ breakInto' 167234520 $ reverse $ sort $ M.keys iSystem
    print $ breakInto 167234520 $ reverse $ sort $ M.keys wSystem
    print $ breakInto 167234520 $ reverse $ sort $ M.keys iSystem

    print $ convert 15
    print $ convert 30
    print $ convert 671

    print $ map (\x -> (convert (fst x), snd x)) $ breakInto 167234520 $ reverse $ sort $ M.keys wSystem
    print $ map (\x -> (convert (fst x), snd x)) $ breakInto 167234520 $ reverse $ sort $ M.keys iSystem